pkg/workspace/plugins.go

```diff
@@ -409,6 +424,80 @@ func GetPluginPath(kind PluginKind, name string, version *semver.Version) (strin
 	return "", "", nil
 }
 
+type sortedPluginInfo []PluginInfo
+
+func (sp sortedPluginInfo) Len() int { return len(sp) }
+func (sp sortedPluginInfo) Less(i, j int) bool {
+	iVersion := sp[i].Version
+	jVersion := sp[j].Version
+	switch {
+	case iVersion == nil && jVersion == nil:
+		return false
+	case iVersion == nil:
+		return true
+	case jVersion == nil:
+		return false
+	default:
+		return iVersion.LT(*jVersion)
+	}
+}
+func (sp sortedPluginInfo) Swap(i, j int) { sp[i], sp[j] = sp[j], sp[i] }
+
+// SelectCompatiblePlugin selects a plugin from the list of plugins with the given kind and name that sastisfies the
+// requested semver range. It returns the highest version plugin that satisfies the requested constraints, or an error
+// if no such plugin could be found.
+//
+// If there exist plugins in the plugin list that don't have a version, SelectCompatiblePlugin will select them if there
+// are no other compatible plugins available.
+func SelectCompatiblePlugin(
+	plugins []PluginInfo, kind PluginKind, name string, requested semver.Range) (PluginInfo, error) {
+	logging.V(7).Infof("SelectCompatiblePlugin(..., %s): beginning", name)
+	var bestMatch PluginInfo
+	var hasMatch bool
+
+	// Before iterating over the list of plugins, sort the list of plugins by version in ascending order. This ensures
+	// that we can do a single pass over the plugin list, from lowest version to greatest version, and be confident that
+	// the best match that we find at the end is the greatest possible compatible version for the requested plugin.
+	sort.Sort(sortedPluginInfo(plugins))
+	for _, plugin := range plugins {
+		switch {
+		case plugin.Kind != kind || plugin.Name != name:
+			// Not the plugin we're looking for.
+			continue
```

Comment:
Nit: we can drop these continues, right? Looks like they're

Reply:
I think so? I'll try it.